src/query/storage/m3/multi_fetch_result.go

```diff
@@ -117,55 +143,84 @@ func (r *multiResult) Add(
 		return
 	}
 
-	r.seenIters = append(r.seenIters, iterators)
+	if len(r.seenIters) == 0 {
+		// store the first attributes seen
+		r.seenFirstAttrs = attrs
+	}
+	r.seenIters = append(r.seenIters, newIterators)
+
+	// Need to check the error to bail early after accumulating the iterators
+	// otherwise when we close the the multi fetch result
 	if !r.err.Empty() {
 		// don't need to do anything if the final result is going to be an error
 		return
 	}
 
-	iters := iterators.Iters()
-	if r.dedupeMap == nil {
-		r.dedupeMap = make(map[string][]multiResultSeries, len(iters))
+	if len(r.seenIters) < 2 {
+		// don't need to create the de-dupe map until we need to actually need to
+		// dedupe between two results
+		return
 	}
 
-	for _, iter := range iters {
-		id := iter.ID().String()
-		r.dedupeMap[id] = append(r.dedupeMap[id], multiResultSeries{
-			attrs: attrs,
-			iter:  iter,
-		})
+	if len(r.seenIters) == 2 {
+		// need to backfill the dedupe map from the first result first
+		existing := r.seenIters[0]
+		r.dedupeMap = make(map[string]multiResultSeries, existing.Len())
+		for _, iter := range existing.Iters() {
```

Comment:
nit: could you refactor this loop into `func (r *multiResult) addOrUpdateMap(attrs, iters)` - would be re-usable here and in `dedupe()`

Reply:
Sure thing.